localstack/plugin/manager.py

```diff
@@ -0,0 +1,316 @@
+    """
+
+    name: str
+    lock: threading.Lock
+
+    plugin_spec: PluginSpec
+    plugin: P = None
+    load_value: Any = None
+
+    is_init: bool = False
+    is_loaded: bool = False
+
+    init_error: Exception = None
+    load_error: Exception = None
+
+
+class PluginManager(PluginLifecycleNotifierMixin, Generic[P]):
+    """
+    Manages Plugins within a namespace discovered by a PluginFinder. The default mechanism is to resolve plugins from
+    entry points using a StevedorePluginFinder.
+    """
+
+    namespace: str
+
+    load_args: Union[List, Tuple]
+    load_kwargs: Dict[str, Any]
+    listener: PluginLifecycleListener
+
+    def __init__(
+        self,
+        namespace: str,
+        load_args: Union[List, Tuple] = None,
+        load_kwargs: Dict = None,
+        listener: PluginLifecycleListener = None,
+        finder: PluginFinder = None,
+    ):
+        self.namespace = namespace
+
+        self.load_args = load_args or list()
+        self.load_kwargs = load_kwargs or dict()
+
+        self.listener = listener or PluginLifecycleListener()
+        self.finder = finder or StevedorePluginFinder(
+            self.namespace, self._fire_on_resolve_exception
+        )
+
+        self._plugin_index = None
+        self._init_mutex = threading.Lock()
+
+    def load(self, name: str) -> P:
+        container = self._require_plugin(name)
+
+        if not container.is_loaded:
+            self._load_plugin(container)
+
+        if container.init_error:
+            raise container.init_error
+
+        if container.load_error:
+            raise container.load_error
+
+        if not container.is_loaded:
+            # plugin should_load returned False
+            raise ValueError("plugin %s:%s is deactivated" % (self.namespace, name))
+
+        return container.plugin
+
+    def load_all(self, propagate_exceptions=False) -> List[P]:
+        plugins = list()
+
+        for name, container in self._plugins.items():
+            if container.is_loaded:
+                plugins.append(container.plugin)
+                continue
+
+            try:
+                plugin = self.load(name)
+                plugins.append(plugin)
+            except Exception:
+                if propagate_exceptions:
+                    raise
+
+        return plugins
+
+    def list_plugin_specs(self) -> List[PluginSpec]:
+        return [container.plugin_spec for container in self._plugins.values()]
+
+    def list_names(self) -> List[str]:
+        return [spec.name for spec in self.list_plugin_specs()]
+
+    def list_containers(self) -> List[PluginContainer[P]]:
+        return list(self._plugins.values())
+
+    def get_container(self, name: str) -> PluginContainer[P]:
+        return self._require_plugin(name)
+
+    def exists(self, name: str) -> bool:
+        return name in self._plugins
+
+    def is_loaded(self, name: str) -> bool:
+        return self._require_plugin(name).is_loaded
```

Comment:
Are all of these convenience functions necessary? (Just asking since they aren't used yet. But on the other hand, there is no more sophisticated implementation yet.)

Reply:
not actually necessary for now, but looked like functions i would use :-) i'll remove those i don't need in the next iteration.